design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,355 @@
+    extract_resource_restrictions:  ExtractResourceRestrictions.new,
+    logger:                         Rails.logger
+  },
+  inputs:    %i(authenticator_name service_id host_name account)
+) do
+
+  def call
+    extract_resource_restrictions_from_annotations
+    extract_resource_restrictions_from_host if resource_restrictions.empty?
+    resource_restrictions
+  end
+
+  private
+  
+  #...
+
+end
+```
+
+This class can be used to replace the `ExtractResourceRestrictions` dependency
+class for k8s (this is command-class composition to mimic inheritance behavior).
+
+#### 2. Validate the Resource Restrictions are defined correctly
+
+In the Azure Authenticator and in the K8s Authenticator we have constraints on 
+how the annotations are defined.
+
+ As can be seen 
+[here](/app/domain/authentication/authn_azure/validate_resource_restrictions.rb), 
+we have the following validations:
+
+  - Validate only permitted restrictions are defined
+    - Here we validate that the `authn-azure` annotations are known. For
+      example, the annotation `authn-azure/subscription-id` is valid, whereas
+      the annotation `authn-azure/unknown` is invalid. We do that to minimize
+      errors from the user, where they will define, for example, the
+      annotation `authn-azure/subscription-ir` (note the typo). In this
+      case we'd rather fail the authentication and point them to their
+      mistake rather than ignore it, and authenticate the request in a weaker
+       granularity than what is expected by the user.
+  - Validate required restrictions exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate mutually exclusive restrictions
+    - For example, in the Azure Authenticator we allow hosts to have
+     only one of the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we allow hosts to have only one of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+
+Also here we can see some generalization.
+
+These need to be defined for each authenticator.
+To simplify that, we can define a simple class to aggregate this information.
+It will allow for each restriction to define its behavior at once, and prevent
+string duplication across the code.
+
+The class will be in the format:
+
+```ruby
+class ResourceRestrictionsConstraints
+
+  attr_reader :permitted, :required, :mutually_exclusive
+
+  def initialize
+    @permitted = []
+    @required = []
+    @mutually_exclusive = {}
+  end
+
+  def add(restriction_name:, required: false, exclusive_group_name: nil)
+    @permitted << restriction_name
+    @required << restriction_name if required
+    (@mutually_exclusive[exclusive_group_name] ||= []) << restriction_name if exclusive_group_name.present?
+    self
+  end
+end
```

Comment:
I like this class dedicated to defining the logic of the resource restrictions.

Really, this is a good example of a value object.  It can be improved by making it immutable.  That is, remove `add` and force the entire thing to be initialized with a single hash defining the three properties.

Than you can add a `validate` method which takes the actual annotation data and validates against the logical definition.  Very similar to what you're proposing here and with the objects below, just a slight difference in the object's API.

Reply:
I removed the `add` as you suggested.
Still, our approach is to put the logic in command classes, and not inside the classes themselves.

Reply:
I finally took @nessiLahav suggestion to create a command-class for each constraint with the same signature.
The command class get the list of configuration in its ctor, and validates it against the Resource Restrictions in the `call`.
This way, each authenticator defines its array of constraints, and the validation only loops through it and calls them.
It allows defining more of these and combining them as necessary in the future.

Reply:
> Still, our approach is to put the logic in command classes, and not inside the classes themselves.

What's your reasoning for wanting to do this?

The problem I see here is that this isn't the intended use of the `CommandClass`.  You don't want to be creating command classes for pure, functional logic, which afaict is what this constraint validation is. 

The only reason you might want a command class here is for the logging: If you want to to log various constraint violations in the same way for all (or multiple) authenticators.  But even there, I think the cleaner design would be to have a pure value object for the constraint logic, which raises errors, and then a very think command class, with the logger, that delegates to the value object and does nothing but map constaint validation errors to log messages.  This keep each object very focused on one responsibility.  Also, the unit tests for the validation logic will be cleaner this way.

Reply:
Thank you @jonahx!
I really appreciate your feedback, and would like to hear more to understand better.

>this isn't the intended use of the `CommandClass`.

So what IS the intended use of the `CommandClass`? Could you elaborate? I would like to know so I could use it better.

Reply:
I changed the constraint classes as you suggested, you can check it out.

Reply:
Cool, I will check it out tomorrow and add feedback.

> So what IS the intended use of the CommandClass? Could you elaborate? I would like to know so I could use it better.

In the meantime, I wrote [this document](https://github.com/cyberark/conjur/blob/master/app/domain/README.md) which tries to answer that question.   This PR might be a good learning experience for me too, re: which parts of that document need to be fleshed out with more examples.  Just glancing over it again, I notice I don't go into how you'd decide when to use a value objects vs a command object, and that's something that would be useful to have some examples for.

Reply:
@abrahamko 

> I changed the constraint classes as you suggested, you can check it out.

The `RequiredConstraint` looks good now, but I liked how you combined them all before too: "all the constraints" seems like a natural way to think of it.  You could achieve that either by combining them into one object like you had before, or by creating an aggregate object around the other three types of objects.  It's hard to say which I'd prefer though without seeing the actual code.

Reply:
I wrapped them in a `MultipleConstraint` object.
How's that?